### PR TITLE
feat!: rewrite as fully-typed go1.27 generic methods

### PR DESCRIPTION
**Pull request by ahmetb**
## Summary

go-linq v5 is a ground-up rewrite built on [Go 1.27 generic methods](https://go.dev/issue/77273). Queries now carry their element type through the entire chain: callbacks receive concrete types, type-changing operators infer their result types at the call site, and the runtime reflection layer is gone entirely. The API keeps v4's operator names and shapes wherever a typed surface allows, so most call sites migrate by deleting their type assertions.

The typed engine is also dramatically faster. On realistic pipelines over 0.5–1M-row datasets, v5 allocates a fixed number of times per query instead of several times per element:

| Pipeline (dataset) | v4 (master) | v5 | Speedup | Allocations (v4 → v5) |
|---|---:|---:|---:|---:|
| `Filter`→`Sum` (1M logs) | 35.3ms | 4.4ms | **8.0×** | 1,499,800 → 3 ❗  |
| `Join` (500k ⋈ 10k) | 26.7ms | 5.4ms | **4.9×** | 1,517,241 → 10,085   ❗ |
| `Select`→`ToSlice` (500k) | 11.2ms | 3.0ms | **3.7×** | 750,098 → 34 ❗  |
| `Distinct` (1M ints) | 23.0ms | 9.3ms | **2.5×** | 997,979 → 535  ❗ |
| `GroupBy` top-5 (1M words) | 37.1ms | 17.6ms | **2.1×** | 1,063,153 → 47,181  ❗ |
| `OrderBy`→`ThenByDesc` (500k) | 230.9ms | 118.0ms | **2.0×** | 26,085,872 → 14  ❗ |
| `Where`→`First` (lazy early-exit) | 42ns | 30ns | 1.4× | 6 → 4 |

Full per-pipeline time and allocation deltas, with a hand-written-loop reference column: [benchmark report](https://floral-knoll-3bqm.here.now/).

**Breaking changes** — documented in [MIGRATION.md](https://github.com/ahmetb/go-linq/blob/v5/MIGRATION.md), point your agent there to migrate. Summary:

- v5 requires Go 1.27.
- Callbacks take concrete types instead of `any`. Migrating a call site usually means deleting its type assertions.
- Terminals that returned `nil` for "not found" (`First`, `Last`, `Single`, `Max`, `Min`, `Aggregate`, …) now return `(value, ok)` pairs.
- `ToSlice` and `ToMap` return their results instead of filling a pointer argument.
- `Sum`, `Average`, `Min`, `Max`, and `ToMap` are now package-level functions (`linq.Sum(q)`); the chainable `SumBy`/`AverageBy`/`MinBy`/`MaxBy`/`ToMapBy` methods cover the fluent style.
- The `…T` reflection method twins and the runtime-sniffing `From(any)` constructor are removed in favor of the typed `From*` constructors.

## Testing Done

The full test suite — operators, edge cases, and doc examples — is ported to the typed API and is green under go1.27rc2, along with vet and gofmt. CI moves to 1.27rc2; lint stays disabled until static-analysis tooling understands generic methods.

The benchmarks above are means of six samples per pipeline (`-benchmem`, go1.27rc2, Apple M5 Pro), and each pipeline's v4 and v5 implementations are parity-tested to return identical results before measuring.

### CODE REVIEW

**Comment by TheFellow** [comment]
We both had the same idea! 😅 #128

**Comment by ahmetb** [comment]
> We both had the same idea! 😅 #128

@TheFellow  lol I'll have it compare the implementations and port improvements from there and let you know how it goes.

File: from.go [hunk 1]
```diff
@@ -2,64 +2,94 @@ package linq
...
-	Value any
+// collect gathers all elements into a slice, preallocating when the query
+// carries a size hint.
+func (q Query[T]) collect() []T {
+	if q.size > 0 {
+		out := make([]T, 0, q.size)
+		q.Iterate(func(item T) bool {
+			out = append(out, item)
+			return true
+		})
```
Comment (kalaninja) [inline from.go]: `MIGRATION.md` currently makes a strong guarantee:

> `ToSlice()` on an empty query returns `nil`

This is not always true when a query carries a stale size hint. Stale hints are explicitly allowed -for example, `FromMap` captures the map length when the query is created, but the map may become empty before it is evaluated. In that case, `collect` allocates a non-nil slice before discovering that the query yields no elements.

It would be better to defer the hinted allocation until the first element is actually yielded:
```suggestion
		out := make([]T, 0, q.size)
		var out []T
		q.Iterate(func(item T) bool {
			if out == nil {
				out = make([]T, 0, q.size)
			}
			out = append(out, item)
			return true
		})
```

You can test it with something like:

```go
func TestToSlice_ReturnsNilWhenSourceMapIsCleared(t *testing.T) {
	source := map[int]int{1: 1}
	q := FromMap(source)
	clear(source)

	if out := q.ToSlice(); out != nil {
		t.Errorf("ToSlice() returned len=%d cap=%d; want nil", len(out), cap(out))
	}
}

File: numeric.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	q.Iterate(func(item T) bool {
+		sum += selector(item)
+		return true
+	})
+	return sum
+}
+
+// Average computes the average of a collection of numeric values.
+// It returns math.NaN() if the collection is empty.
+func Average[TNumber Number](q Query[TNumber]) float64 {
```
Comment (kalaninja) [inline numeric.go]: tbh, I don't like it very much that this is not chained any more, but I undestand the reason.

The way it is implemented now breaks compatibility with v4 and original linq, by converting into float64 too early. E.g. consider the following test

```
func TestAverage_LargeIntegerCancellation(t *testing.T) {
	const float64PrecisionLimit = int64(1 << 53)
	input := []int64{float64PrecisionLimit + 1, -float64PrecisionLimit}
	if got := Average(FromSlice(input)); got != 0.5 {
		t.Errorf("Average(%v)=%v; want 0.5", input, got)
	}
}
```

 I might suggest to introduce a private function
```
func averageBy[T any, TNumber Number](q Query[T], selector func(T) TNumber) float64 {
	var sum float64
	n := 0

	switch kindOfNumber[TNumber]() {
	case signedNumber:
		var integerSum int64
		q.Iterate(func(item T) bool {
			integerSum += int64(selector(item))
			n++
			return true
		})
		sum = float64(integerSum)
	case unsignedNumber:
		var integerSum uint64
		q.Iterate(func(item T) bool {
			integerSum += uint64(selector(item))
			n++
			return true
		})
		sum = float64(integerSum)
	case floatingPointNumber:
		q.Iterate(func(item T) bool {
			sum += float64(selector(item))
			n++
			return true
		})
	}

	if n == 0 {
		return math.NaN()
	}
	return sum / float64(n)
}
```

and delegate other methods to it
```
func Average[TNumber Number](q Query[TNumber]) float64 {
	return averageBy(q, func(item TNumber) TNumber { return item })
}

func (q Query[T]) AverageBy[TNumber Number](selector func(T) TNumber) float64 {
	return averageBy(q, selector)
}
```

File: take.go
Comment (kalaninja) [inline take.go:23]: There’s a bug here: first retrieve an item from the iterator, then decide whether it is needed.

```suggestion
func (q Query[T]) Take(count int) Query[T] {
	if count <= 0 {
		return Query[T]{Iterate: func(func(T) bool) {}}
	}

	return Query[T]{
		Iterate: func(yield func(T) bool) {
			n := count
			q.Iterate(func(item T) bool {
				if !yield(item) {
					return false
				}
				n--
				return n > 0
			})
		},
	}
}
```

Here is a test for it
```
func TestTakePullsExactlyCount(t *testing.T) {
	tests := []struct {
		count  int
		output []int
		pulled int
	}{
		{0, nil, 0},
		{1, []int{1}, 1},
		{3, []int{1, 2, 3}, 3},
		{4, []int{1, 2, 3, 4}, 4},
		{9, []int{1, 2, 3, 4}, 4},
	}

	for _, test := range tests {
		pulled := 0
		q := FromSlice([]int{1, 2, 3, 4}).Where(func(int) bool {
			pulled++
			return true
		}).Take(test.count)

		if out := toSlice(q); !slices.Equal(out, test.output) {
			t.Errorf("Take(%d)=%v expected %v", test.count, out, test.output)
		}
		if pulled != test.pulled {
			t.Errorf("Take(%d) pulled %d elements from the source, expected %d",
				test.count, pulled, test.pulled)
		}
	}
}
```

File: concat.go [hunk 3]
```diff
@@ -41,24 +41,22 @@ func (q Query) Concat(q2 Query) Query {
...
+	return Query[T]{
+		Iterate: func(yield func(T) bool) {
 			if !yield(item) {
 				return
 			}
 
-			q.Iterate(func(item any) bool {
+			q.Iterate(func(item T) bool {
 				return yield(item)
 			})
```
Comment (kalaninja) [inline concat.go]:
```suggestion
			q.Iterate(yield)
```

File: from.go [hunk 4]
```diff
@@ -105,98 +135,45 @@ func FromString[S ~string](source S) Query {
 
 // FromIterable initializes a linq query with a custom collection passed. This
 // collection has to implement Iterable.
-func FromIterable(source Iterable) Query {
-	return Query{
+func FromIterable[T any](source Iterable[T]) Query[T] {
```
Comment (kalaninja) [inline from.go]: I can suggest to remove this method and `type Iterable[T any] interface` in favor of FromSeq you introduced.

File: result.go [hunk 5]
```diff
@@ -261,409 +129,177 @@ func (q Query) ForEachT(actionFn any) {
...
+	})
+	return r, found
 }
 
-// Results collects all items from a query into a slice.
-func (q Query) Results() []any {
-	return slices.Collect(q.Iterate)
+// Results collects all items from a query into a slice. It is equivalent to
+// ToSlice and is kept for familiarity with earlier go-linq versions.
+func (q Query[T]) Results() []T {
```
Comment (kalaninja) [inline result.go]: I think we should remove this since we have a proper ToSlice method now

File: set.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	case complex128:
+		f = typedEqual[complex128]()
+	case string:
+		f = typedEqual[string]()
+	case bool:
+		f = typedEqual[bool]()
+	default:
+		return func(a, b T) bool { return any(a) == any(b) }
+	}
+	return f.(func(T, T) bool)
```
Comment (kalaninja) [inline set.go]: looks like the same switch is repeated in newSeenSet too. It can probably combined similar to:

```
type ops[T any] struct {
	newSet func() seenSet[T]
	equal  func(T, T) bool
}

func basicOps[T comparable]() any {
	return ops[T]{
		newSet: func() seenSet[T] { return typedSet[T]() },
		equal:  func(a, b T) bool { return a == b },
	}
}

func opsFor[T any]() ops[T] {
	var o any
	switch any(*new(T)).(type) {
	case int:
		o = basicOps[int]()
	case int8:
		o = basicOps[int8]()
	case int16:
		o = basicOps[int16]()
	case int32:
		o = basicOps[int32]()
	case int64:
		o = basicOps[int64]()
	case uint:
		o = basicOps[uint]()
	case uint8:
		o = basicOps[uint8]()
	case uint16:
		o = basicOps[uint16]()
	case uint32:
		o = basicOps[uint32]()
	case uint64:
		o = basicOps[uint64]()
	case uintptr:
		o = basicOps[uintptr]()
	case float32:
		o = basicOps[float32]()
	case float64:
		o = basicOps[float64]()
	case complex64:
		o = basicOps[complex64]()
	case complex128:
		o = basicOps[complex128]()
	case string:
		o = basicOps[string]()
	case bool:
		o = basicOps[bool]()
	default:
		return ops[T]{
			newSet: boxedSet[T],
			equal:  func(a, b T) bool { return any(a) == any(b) },
		}
	}
	return o.(ops[T])
}

func newSeenSet[T any]() seenSet[T] { return opsFor[T]().newSet() }

func equalFor[T any]() func(T, T) bool { return opsFor[T]().equal }
```

I can also see that equalFor is needed as a comparator in some methods (e.g. Contains) as an alternative you can probably split it into 2 methods like you did for Average() and AverageBy():

```
func Contains[T comparable](q Query[T], value T) bool

func (q Query[T]) ContainsBy[TKey comparable](sel func(T) TKey, key TKey) bool
```
Not great, but we already have it in API and it might allow to get rid of equalFor.

File: take.go
Comment (ahmetb) [reply take.go:23]: That's a really good catch. Fixed and added a test case for it: [05d452d](https://github.com/ahmetb/go-linq/pull/129/commits/05d452d854c544f43b11605f7430efeb57337689)

File: numeric.go (hunk 2 above)
Comment (ahmetb) [reply numeric.go]: Done. b60f1ce 
This also added 10% perf improvement.

File: result.go (hunk 5 above)
Comment (ahmetb) [reply result.go]: Done in 0b7ba06

File: set.go (hunk 6 above)
Comment (ahmetb) [reply set.go]: Done in 15b9955.

**Comment by cleitonmarx** [comment]
It looks great!

File: orderby.go [hunk 7]
```diff
@@ -1,282 +1,157 @@
...
+//
+// Unlike OrderBy, Sort does not require the sort key to be an ordered type,
+// so it can be used with arbitrary comparison logic. The less function is
+// invoked once per element comparison.
+func (q Query[T]) Sort(less func(i, j T) bool) Query[T] {
+	return Query[T]{
+		Iterate: func(yield func(T) bool) {
+			items := q.collect()
+
+			sort.Sort(sorter[T]{items: items, less: less})
```
Comment (kalaninja) [inline orderby.go:147]: I think we might reuse newly introduced `sortedIterate` here and remove `type sorter[T any] struct` completely.

e.g. if you prefer to keep function signature.
```
func (q Query[T]) Sort(less func(i, j T) bool) Query[T] {
	compare := func(a, b T) int {
		if less(a, b) {
			return -1
		}
		return 1
	}

	return Query[T]{
		Iterate: q.sortedIterate([]func(a, b T) int{compare}),
		size:    q.size,
	}
}
```

but I'd probably even changed it for something better looking like:
```
func (q Query[T]) Sort(compare func(a, b T) int) Query[T] {
	return Query[T]{
		Iterate: q.sortedIterate([]func(a, b T) int{compare}),
		size:    q.size,
	}
}
```

File: numeric.go [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+func averageFloat[T any, TNumber Number](q Query[T], selector func(T) TNumber) (float64, int) {
+	var sum float64
+	n := 0
+	q.Iterate(func(item T) bool {
+		sum += float64(selector(item))
+		n++
+		return true
+	})
+	return sum, n
+}
```
Comment (kalaninja) [inline numeric.go:105]: we can probably substitute it with generic:

```suggestion
func averageAs[Acc int64 | uint64 | float64, T any, TNumber Number](
	q Query[T], selector func(T) TNumber) (float64, int) {
	var sum Acc
	n := 0
	q.Iterate(func(item T) bool {
		sum += Acc(selector(item))
		n++
		return true
	})
	return float64(sum), n
}
```

and call like
```
	switch kindOfNumber[TNumber]() {
	case signedNumber:
		sum, n = averageAs[int64](q, selector)
	case unsignedNumber:
		sum, n = averageAs[uint64](q, selector)
	default: // floatingPointNumber
		sum, n = averageAs[float64](q, selector)
	}
```

File: result.go [hunk 9]
```diff
@@ -261,409 +129,171 @@ func (q Query) ForEachT(actionFn any) {
...
-	}
+		return true
+	})
 
-	return
+	if !visited || !ok {
+		return zero, false
+	}
+	return r, true
 }
```
Comment (kalaninja) [inline result.go:221]: this might be done with a counter instead of 2 bools (should be faster)


```suggestion
func (q Query[T]) Single() (T, bool) {
	var r T
	count := 0

	q.Iterate(func(item T) bool {
		count++
		if count == 2 {
			return false
		}
		r = item
		return true
	})

	if count == 1 {
		return r, true
	}
	var zero T
	return zero, false
}
```

File: result.go [hunk 10]
```diff
@@ -261,409 +129,171 @@ func (q Query) ForEachT(actionFn any) {
...
 
 // SingleWith returns the only element of a collection that satisfies a
-// specified condition, and nil if more than one such element exists.
-func (q Query) SingleWith(predicate func(any) bool) (r any) {
+// specified condition and a boolean that is true only if exactly one such
+// element exists.
+func (q Query[T]) SingleWith(predicate func(T) bool) (T, bool) {
```
Comment (kalaninja) [inline result.go:226]: same idea as Single() above

```
func (q Query[T]) SingleWith(predicate func(T) bool) (T, bool) {
	var r T
	count := 0

	q.Iterate(func(item T) bool {
		if !predicate(item) {
			return true
		}
		count++
		if count == 2 {
			return false
		}
		r = item
		return true
	})

	if count == 1 {
		return r, true
	}
	var zero T
	return zero, false
}
```

File: orderby.go [hunk 11]
```diff
@@ -1,282 +1,157 @@
...
-				items := q.sort([]order{{selector: selector, desc: true}})
-				for _, item := range items {
-					if !yield(item) {
-						return
-					}
-				}
-			},
+		Query: Query[T]{
+			Iterate: q.sortedIterate(compares),
+			size:    q.size,
```
Comment (kalaninja) [inline orderby.go:72]: this part here and the next 3 methods repeat the same code. we can reduce to

```
func newOrderedQuery[T any](original Query[T], compares []func(a, b T) int) OrderedQuery[T] {
	return OrderedQuery[T]{
		compares: compares,
		original: original,
		Query: Query[T]{
			Iterate: original.sortedIterate(compares),
			size:    original.size,
		},
	}
}

func (q Query[T]) OrderBy[TKey cmp.Ordered](selector func(T) TKey) OrderedQuery[T] {
	return newOrderedQuery(q, []func(a, b T) int{ascending(selector)})
}

func (q Query[T]) OrderByDescending[TKey cmp.Ordered](selector func(T) TKey) OrderedQuery[T] {
	return newOrderedQuery(q, []func(a, b T) int{descending(selector)})
}

func (oq OrderedQuery[T]) ThenBy[TKey cmp.Ordered](selector func(T) TKey) OrderedQuery[T] {
	compares := append(slices.Clip(oq.compares), ascending(selector))
	return newOrderedQuery(oq.original, compares)
}

func (oq OrderedQuery[T]) ThenByDescending[TKey cmp.Ordered](selector func(T) TKey) OrderedQuery[T] {
	compares := append(slices.Clip(oq.compares), descending(selector))
	return newOrderedQuery(oq.original, compares)
}
```

File: from.go [hunk 12]
```diff
@@ -2,64 +2,93 @@ package linq
...
-		Iterate: func(yield func(any) bool) {
+func FromSlice[S ~[]T, T any](source S) Query[T] {
+	return Query[T]{
+		Iterate: func(yield func(T) bool) {
 			for _, item := range source {
 				if !yield(item) {
 					return
 				}
 			}
 		},
```
Comment (kalaninja) [inline from.go:64]:
```suggestion
		Iterate: slices.Values(source),),
```

File: defaultifempty.go [hunk 13]
```diff
@@ -2,13 +2,13 @@ package linq
 
 // DefaultIfEmpty returns the elements of the specified sequence
 // if the sequence is empty.
-func (q Query) DefaultIfEmpty(defaultValue any) Query {
-	return Query{
-		Iterate: func(yield func(any) bool) {
+func (q Query[T]) DefaultIfEmpty(defaultValue T) Query[T] {
```
Comment (kalaninja) [inline defaultifempty.go:5]: this can probably be simplified
```
func (q Query[T]) DefaultIfEmpty(defaultValue T) Query[T] {
	return Query[T]{
		Iterate: func(yield func(T) bool) {
			empty := true

			q.Iterate(func(item T) bool {
				empty = false
				return yield(item)
			})

			if empty {
				yield(defaultValue)
			}
		},
	}
}
```

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -14,20 +15,10 @@ env:
   GO111MODULE: on
```
Comment (kalaninja) [inline .github/workflows/ci.yml:15]: this is probably not needed any more since we are targeting 1.27?

**Comment by kalaninja** [comment]
@ahmetb I opened #132 as a follow-up proposal for v5. It improves size-hint propagation for operators whose output cardinality can be derived exactly, enabling more accurate preallocation without changing query semantics.

**Comment by ahmetb** [comment]
@kalaninja I also had a branch about size-hinting based perf improvements in https://github.com/ahmetb/go-linq/tree/ahmet/size-hints (forgot to push this up earlier, feel free to incorporate it into #132). 

```
┌────────────────────────────────────────┬────────────────────┬──────────────────────────────────────────┬────────┬───────────┐
│                 Change                 │     Benchmark      │                   Time                   │ Memory │  Allocs   │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ Size hints through Take                │ Take(500k)→ToSlice │ −33%                                     │ −81%   │ 40→5      │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ …through Concat                        │ 1M+100k→ToSlice    │ −31%                                     │ −83%   │ 44→5      │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ …through DefaultIfEmpty                │ 1M→ToSlice         │ −29%                                     │ −81%   │ 43→5      │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ …through Append/Prepend                │ 1M→ToSlice         │ −18%                                     │ −81%   │ 43→5      │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ …through Skip                          │ Skip(500k)→ToSlice │ −16%                                     │ −81%   │ 40→5      │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ …through Zip                           │ 1M pairs→ToSlice   │ −3%                                      │ −81%   │ 49→11     │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ Join lookup presized from inner's hint │ 100k⋈10k           │ −22%                                     │ −47%   │ ~         │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ ToMap/ToMapBy presized from hint       │ 1M entries         │ −14%                                     │ −50%   │ 8.2k→4.1k │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ Union/UnionBy single emit closure      │ 1M∪100k            │ ~                                        │ ~      │ −1 fixed  │
├────────────────────────────────────────┼────────────────────┼──────────────────────────────────────────┼────────┼───────────┤
│ FromSlice via slices.Values            │ Select/Where/Sum   │ neutral — kept for simplicity (−6 lines) │        │           │
└────────────────────────────────────────┴────────────────────┴──────────────────────────────────────────┴────────┴───────────┘
```
